### PR TITLE
[Merged by Bors] - cleanups around checkpoint-recovery

### PR DESCRIPTION
**Pull request by poszu**
## Motivation

few refactors to simplify and untangle the code

## Description

- cleaned up unit tests:
  - removed implicit usage of globals (`expectedCheckpoint`)
  - reordered tests to avoid tests branching on conditions like `if tc.fail` (tests for error conditions are separate now)
- moved loading checkpoint out from the Cobra command to the app start routine
- removed `preserve *checkpoint.PreserveData` from `App struct` - it's a local variable now
- moved few checks from `node` package to `checkpoint` for better testability + added tests

## Test Plan

tests pass

## TODO

- [x] Explain motivation or link existing issue(s)
- [x] Test changes and document test plan
- [ ] Update documentation as needed
- [ ] Update [changelog](../CHANGELOG.md) as needed

### CODE REVIEW

File: checkpoint/recovery.go [hunk 1]
```diff
@@ -108,6 +108,13 @@ func Recover(
 	fs afero.Fs,
 	cfg *RecoverConfig,
 ) (*PreservedData, error) {
+	if len(cfg.Uri) == 0 {
+		return nil, nil
```
Comment (acud) [inline checkpoint/recovery.go]: NIT: I think that methods that return `(*pointer,err)` shouldn't `return nil, nil` as it can cause panics on the calling context. In such cases my tendency would be to add another `bool` value that indicates whether something was done or not for clearer flow control on the calling context (or alternatively, since `cfg` is passed from somewhere, that check could be done on the calling context).

File: checkpoint/runner_test.go [hunk 2]
```diff
@@ -231,89 +252,68 @@ func asAtxSnapshot(v *types.ActivationTx, cmt *types.ATXID) types.AtxSnapshot {
...
-func createMesh(t *testing.T, db *sql.Database, miners map[types.NodeID][]*types.ActivationTx, accts []*types.Account) {
-	for _, vatxs := range miners {
-		vatxs = slices.Clone(vatxs)
-		// ATXs are expected to be in reverse epoch order and we want older ATXs
-		// created first so that the nonce can be retrieved from them for
-		// populating the nonce field when creating newer ones
-		slices.Reverse(vatxs)
-		for _, atx := range vatxs {
+func createMesh(t *testing.T, db *sql.Database, miners []miner, accts []*types.Account) {
+	for _, miner := range miners {
```
Comment (acud) [inline checkpoint/runner_test.go:257]: Add `t.Helper()` before this line

File: node/node.go [hunk 3]
```diff
@@ -2169,40 +2157,32 @@ func (app *App) startSynchronous(ctx context.Context) (err error) {
 	return nil
 }
 
-func (app *App) preserveAfterRecovery(ctx context.Context) {
-	if app.preserve == nil {
+func (app *App) preserveAfterRecovery(ctx context.Context, preserved *checkpoint.PreservedData) {
+	if preserved == nil {
```
Comment (acud) [inline node/node.go]: NIT: since this is already known before the call to this method, maybe we could just do this check before `L#2144`?

Also, this method seem to be able to just error continuously without any consequence - all errors are swallowed and the execution continues. My suggestion/claim is that if an error is swallowed and is recoverable then it is not really an error. So in this case I'm doubting whether these errors should be: 1) logged as `Error`? (maybe just on Debug/Warn) 2) if they are recoverable is it of any use to [display to the end user](https://github.com/spacemeshos/go-spacemesh/blob/develop/CODING.md#users)? 3) logged at all?
This comment is a ditto for all the errors below which are swallowed.

**Review by acud — CHANGES_REQUESTED** [review]
:pray: A few comments from my side

File: checkpoint/recovery.go (hunk 1 above)
Comment (fasmat) [reply checkpoint/recovery.go]: I'm not sure that returning no error here is even correct? 🤔 shouldn't it return an error if no URI was passed?

File: checkpoint/runner_test.go (hunk 2 above)
Comment (fasmat) [reply checkpoint/runner_test.go:257]: Probably also should be `testing.TB` instead of `*testing.T` 🙂

File: checkpoint/runner_test.go [hunk 4]
```diff
@@ -231,89 +252,68 @@ func asAtxSnapshot(v *types.ActivationTx, cmt *types.ATXID) types.AtxSnapshot {
...
-		{
-			desc:  "no atxs",
-			accts: allAccounts,
-			fail:  true,
-		},
-		{
-			desc:  "no accounts",
-			atxes: allAtxs,
-			fail:  true,
-		},
```
Comment (fasmat) [inline checkpoint/runner_test.go:292]: I'm not sure if we should remove test cases with invalid data checking if they fail as expected?

Comment (poszu) [reply checkpoint/runner_test.go:292]: They are not removed, but moved below

File: checkpoint/recovery.go (hunk 1 above)
Comment (poszu) [reply checkpoint/recovery.go]: I intentionally moved this check from node.go to be able to test it (testing the `App` is unpleasant). Empty URI essentially stands for "checkpoint-restore" disabled and it's a valid value, hence returning `nil` error.

> I think that methods that return (*pointer,err) shouldn't return nil, nil as it can cause panics on the calling context.

Well, it _depends_. Unfortunately, Go doesn't have something like Rust `Option` and we need to escape to pointers with `nil` equivalent to `None` :(. I could document the possible return values if that would help, although I'd prefer to come up with an API that's hard to misuse rather than depend on comments which are rarely read.

File: checkpoint/runner_test.go (hunk 2 above)
Comment (poszu) [reply checkpoint/runner_test.go:257]: > Probably also should be `testing.TB` instead of `*testing.T` 🙂

Why? We rather use the latter in most places in the codebase. TBH either is fine for me as long as we are consistent throughout the repo.

File: checkpoint/recovery.go (hunk 1 above)
Comment (fasmat) [reply checkpoint/recovery.go]: The caller could just check if `URI` is not set and then not call the function? as the function would return an error otherwise.

File: checkpoint/runner_test.go (hunk 2 above)
Comment (fasmat) [reply checkpoint/runner_test.go:257]: `testing.TB` is an interface and `*testing.T` implements that interface, but also `*testing.B`.

> We rather use the latter in most places in the codebase.

I would say that this is not true. Helper functions often take `testing.TB` (as they should) and even in your PR you use `testing.TB` a lot, just see `checkpoint/recovery_test.go`.

Generally test helper methods should in my opinion use the interface rather than the concrete type, so they can be used in both tests and benchmarks.

File: checkpoint/recovery.go (hunk 1 above)
Comment (poszu) [reply checkpoint/recovery.go]: > The caller could just check if `URI` is not set and then not call the function? as the function would return an error otherwise.

Then how do you propose to test that an empty URI is a valid node configuration value?

Comment (fasmat) [reply checkpoint/recovery.go]: If its empty it won't call the `Recover` function? A negative doesn't need to be tested?

Comment (fasmat) [reply checkpoint/recovery.go]: It's also indirectly tested by every other test in the `node` package which would fail if `Recover` would do something on a non-empty `URI` 😉

File: checkpoint/runner_test.go (hunk 2 above)
Comment (poszu) [reply checkpoint/runner_test.go:257]: > I would say that this is not true. Helper functions often take testing.TB (as they should) 

I counted. There are ~104~ 148 helper functions using `*testing.T` and ~12~ 160 using `testing.TB`. Regexps used:

```
rg '\*testing.T\b' | rg -v 'func\(t \*testing\.T\b' | rg -v 'func Test.*\(t \*testing\.T\)' | wc -l
rg 'testing.TB\b' | wc -l
```


> and even in your PR you use testing.TB a lot, just see checkpoint/recovery_test.go.

I didn't add any new helper method, they are using what they were created with :shrug:.

> Generally test helper methods should in my opinion use the interface rather than the concrete type, so they can be used in both tests and benchmarks.

I'm fine with that but let's be consistent.

Comment (fasmat) [reply checkpoint/runner_test.go:257]: You also removed `testing.TB` from a few methods in this PR 😉 

I'm advocating for using `testing.TB` because you can use the helper then in both tests and benchmarks and "other places don't do that" imo isn't an argument for not doing that?  😕

File: node/node.go (hunk 3 above)
Comment (poszu) [reply node/node.go]: >NIT: since this is already known before the call to this method, maybe we could just do this check before L#2144?

:+1: Moved the check there.

Regarding error handling in this function. I'm not 100% sure. I think the author's intent was that this is "best-effort". But that doesn't make much sense because a node that failed to recover (preserve) its own dependencies is unlikely to be able to function properly. So I think it could very well just terminate. @fasmat WDYT?

File: checkpoint/runner_test.go (hunk 2 above)
Comment (poszu) [reply checkpoint/runner_test.go:257]: > You also removed `testing.TB` from a few methods in this PR 😉

I removed them from where they were unused.

> "other places don't do that" imo isn't an argument for not doing that? 😕

I'm not saying that it is :shrug:. I'm saying "let's be consistent".

**Review by fasmat — APPROVED** [review]

**Review by acud — APPROVED** [review]
LGTM @poszu, thanks!

**Comment by poszu** [comment]
Bors merge

**Comment by poszu** [comment]
git on windows failed

bors merge

File: node/node.go (hunk 3 above)
Comment (fasmat) [reply node/node.go]: The `preserve` is the data the local node collected it knew about, but wasn't contained in the checkpoint. Failing to restore it won't put the node in an unusable state (the checkpoint has successfully been restored). But data of the node that might be important for it (and wasn't included in the checkpoint) is now gone.

Afaik there is no way to recover from this (unless the node has a backup) but the node should still work normally.

**Comment by poszu** [comment]
Flaky TestPartition

bors merge

**Comment by poszu** [comment]
bors merge

**Comment by poszu** [comment]
bors merge
